src/core/cm/updatemanager/unitstatushandler.cpp

```diff
@@ -0,0 +1,472 @@
+        return AOS_ERROR_WRAP(err);
+    }
+
+    if (auto err = SetInstancesStatus(); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    if (auto err = SetUnitSubjects(); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    LogUnitStatus();
+
+    // Send unit status
+
+    if (auto err = mSender->SendUnitStatus(mUnitStatus); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    ClearUnitStatus();
+
+    return ErrorEnum::eNone;
+}
+
+void UnitStatusHandler::SetCloudConnected(bool isConnected)
+{
+    LockGuard lock {mMutex};
+
+    mCloudConnected = isConnected;
+
+    if (!mCloudConnected) {
+        mTimer.Stop();
+    }
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+void UnitStatusHandler::OnNodeInfoChanged(const UnitNodeInfo& info)
+{
+    LockGuard lock {mMutex};
+
+    LOG_INF() << "Node info changed" << Log::Field("id", info.mNodeID) << Log::Field("type", info.mNodeType)
+              << Log::Field("provisioned", info.mProvisioned) << Log::Field("state", info.mState)
+              << Log::Field(info.mError);
+
+    if (!mCloudConnected) {
+        return;
+    }
+
+    if (!mUnitStatus.mNodes.HasValue()) {
+        mUnitStatus.mNodes.EmplaceValue();
+    }
+
+    auto it = mUnitStatus.mNodes->FindIf(
+        [&info](const UnitNodeInfo& nodeInfo) { return nodeInfo.mNodeID == info.mNodeID; });
+    if (it != mUnitStatus.mNodes->end()) {
+        *it = info;
+    } else {
+        if (auto err = mUnitStatus.mNodes->EmplaceBack(info); !err.IsNone()) {
+            LOG_ERR() << "Failed to emplace node info" << Log::Field(err);
+            return;
+        }
+    }
+
+    StartTimer();
+}
+
+void UnitStatusHandler::OnImageStatusChanged(const String& itemID, const String& version, const ImageStatus& status)
+{
+    LockGuard lock {mMutex};
+
+    LOG_INF() << "Image status changed" << Log::Field("itemID", itemID) << Log::Field("version", version)
+              << Log::Field("imageID", status.mImageID) << Log::Field("state", status.mState)
+              << Log::Field(status.mError);
+
+    if (!mCloudConnected) {
+        return;
+    }
+
+    if (!mUnitStatus.mUpdateItems.HasValue()) {
+        mUnitStatus.mUpdateItems.EmplaceValue();
+    }
+
+    auto itemIt = mUnitStatus.mUpdateItems->FindIf([&itemID, &version](const UpdateItemStatus& itemStatus) {
+        return itemStatus.mItemID == itemID && itemStatus.mVersion == version;
+    });
+    if (itemIt == mUnitStatus.mUpdateItems->end()) {
+        if (auto err = mUnitStatus.mUpdateItems->EmplaceBack(itemID, version); !err.IsNone()) {
+            LOG_ERR() << "Failed to emplace update item status" << Log::Field(err);
+            return;
+        }
+
+        itemIt = mUnitStatus.mUpdateItems->end() - 1;
```

Comment:
This way looks less magic :)
```suggestion
        itemIt = &mUnitStatus.mUpdateItems->Back();
```

Reply:
agree, fixed